From 124f67f642f031e7afbcebe4b1ea8bdceafebb96 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sat, 31 Mar 2018 13:14:11 +0200
Subject: [PATCH 1/1] efi_selftest: test unaligned memory access

According to the UEFI spec unaligned memory access should be enabled on
CPUs supporting it.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_selftest/Makefile                 |  4 +++
 lib/efi_selftest/efi_selftest_unaligned.c | 57 +++++++++++++++++++++++++++++++
 2 files changed, 61 insertions(+)
 create mode 100644 lib/efi_selftest/efi_selftest_unaligned.c

diff --git a/lib/efi_selftest/Makefile b/lib/efi_selftest/Makefile
index 31b444fc8b..2ca58cebe4 100644
--- a/lib/efi_selftest/Makefile
+++ b/lib/efi_selftest/Makefile
@@ -31,6 +31,10 @@ efi_selftest_tpl.o \
 efi_selftest_util.o \
 efi_selftest_watchdog.o
 
+ifeq ($(CONFIG_CMD_BOOTEFI_SELFTEST),y)
+obj-$(CONFIG_CPU_V7) += efi_selftest_unaligned.o
+endif
+
 ifeq ($(CONFIG_BLK)$(CONFIG_PARTITIONS),yy)
 obj-$(CONFIG_CMD_BOOTEFI_SELFTEST) += efi_selftest_block_device.o
 endif
diff --git a/lib/efi_selftest/efi_selftest_unaligned.c b/lib/efi_selftest/efi_selftest_unaligned.c
new file mode 100644
index 0000000000..6b81fc8a0b
--- /dev/null
+++ b/lib/efi_selftest/efi_selftest_unaligned.c
@@ -0,0 +1,57 @@
+/*
+ * efi_selftest_unaligned
+ *
+ * Copyright (c) 2018 Heinrich Schuchardt <xypron.glpk@gmx.de>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Test unaligned memory access
+ */
+
+#include <efi_selftest.h>
+
+struct aligned_buffer {
+	char a[8] __aligned(8);
+};
+
+static inline u32 deref(u32 *addr)
+{
+	int ret;
+	asm(
+		"ldr %[out], [%[in], #1]\n\t"
+		: [out] "=r" (ret)
+		: [in]"r" (addr)
+		:
+	);
+	return ret;
+}
+
+/*
+ * Execute unit test.
+ *
+ * @return:	EFI_ST_SUCCESS for success
+ */
+static int execute(void)
+{
+	struct aligned_buffer buf = {
+		{0, 1, 2, 3, 4, 5, 6, 7},
+		};
+	void *v = &buf;
+	u32 r = 0;
+
+	r = deref(v);
+
+	/* UEFI only supports low endian systems */
+	if (r != 0x04030201) {
+		efi_st_error("Unaligned access failed");
+		return EFI_ST_FAILURE;
+	}
+
+	return EFI_ST_SUCCESS;
+}
+
+EFI_UNIT_TEST(unaligned) = {
+	.name = "unaligned memory access",
+	.phase = EFI_EXECUTE_BEFORE_BOOTTIME_EXIT,
+	.execute = execute,
+};
-- 
2.14.2

